### ISSUE TITLE
align left and right on the same line

### ISSUE THREAD

**Issue by GoogleCodeExporter** [python-escpos/python-escpos#27]
```
What steps will reproduce the problem?
1. use Epson TM-T20
2. execute the following code
3.

What is the expected output? What do you see instead?
I would like to format a line to begin on the left with some words ("article" 
in my example) and then align on the right to write the price (15.00 € here)

What version of the product are you using? On what operating system?
python-escpos 1.0-1 on ubuntu 13.10

Please provide any additional information below.

code :
Epson.set("LEFT", "A", "", 1, 1)
Epson.text("Article 1")
Epson.set("RIGHT", "A", "", 1, 1)
Epson.text("15.00 \xd5")
Epson.text("\n")

```

Original issue reported on code.google.com by `shu...@gmail.com` on 15 May 2014 at 10:29

**Comment by patkan** [comment]
I don't think that escpos printers are capable of this. Probably the best way to do that is by using tabs.

**Comment by MicroJoe** [comment]
I have the same needs as in this issue but I don't know at all how tabs work. Any link to some documentation or something about setting tabs for a printer?

**Comment by miracle2k** [comment]
The way I am doing this myself is, using knowledge of the column count, inserting the required amount of whitespace.

I am using https://github.com/miracle2k/py-xml-escpos.

**Comment by patkan** [comment]
@MicroJoe For Epson there is the website at https://reference.epson-biz.com/modules/ref_escpos/index.php?content_id=2
If you create a login it also has some nice structuring. Maybe this can help?

**Comment by MicroJoe** [comment]
I have the WinBond/Zjiang 5890 on my desk. Given the problems I encounter with this printer it may handle the horizontal tabs wrong.

I think the best option is to use the printer knowledge database in order to retrieve the column number as explained by @miracle2k and then fill with spaces. I think it should be integrated in the libràry (maybe we can call this "software horizontal tabs support").

**Comment by patkan** [comment]
@MicroJoe this sounds sensible to me.
Probably best would be getting #161 merged (this is probably my task) and then building on top of this the "software" tabs. (Would you like to do this?)

**Comment by MicroJoe** [comment]
Yeah I think #161 should be merged before. Then we can add a `softtab=False` default attribute that would store the tab positions in an array and then when `text` is called we can match the `\t` characters and use spaces to layout text.

We could even add more capabilites to the software tabs like align left or right on tabs, but this will then make software tabs incompatible with hardware tabs (not sure about this because at the moment I don't understand at all how hardware horizontal tabs work, someone should explain me this before I implement something).

**Comment by patkan** [comment]
I am also not 100% how hardware tabs work. But as far as I get it, you can set multiples of columns as tab positions before printing. (Iirc you can have 16) Then, during printing, you can jump to the next horizontal tab with the proper command.

**Comment by MicroJoe** [comment]
Ok, I have played with hardware tabs on my printer with the following snippet:
```python
for i in range(0,32):
    p.control('HT', i)
    p.text('test\tff\n')

p.print_and_feed(2)
```
Of course I had to patch the 1 to 16 limit in the control method:
```python
        elif ctl.upper() == "HT":
            #if not (1 <= pos <= 16):
            #    raise TabPosError()
            #else:
                # Set tab positions
            self._raw(CTL_SET_HT + six.int2byte(pos))
```
And I got this result:
![tabtest](https://cloud.githubusercontent.com/assets/989521/26074905/c48c554c-39b3-11e7-8e71-ee569f94bab1.jpg)

So it seems to me that the `control('HT', x)` method sets the **tab size** and not the tab position. I am not sure if this is because of my printer or if this behavior is observed on reference Epson hardware. If this behavior is the same on Epson printers we can create a `set_tab_size` method or similar, and remove the 1-16 limit.

**Comment by miracle2k** [comment]
Unfortunately I never got around to implementing  it, but if we were to separate the API into a low-level ESC/POS wrapper and a higher-level layout API (https://github.com/python-escpos/python-escpos/issues/176), we could implement soft-tabs in the latter, while the raw tabs API is separate. In other words, it needn't worry about compatibility between soft and hardtabs if for the user the choice is abstracted away in the layout system.

Having said that, I wonder in what way hard tabs might be superior to soft tabs. Is there anything lost by using softtabs exclusively? Larger data sizes, maybe?

**Comment by MicroJoe** [comment]
I had to use this feature for receipt-like printing recently. Here is the snippet that I used (with hardcoded width from my printer):

```python
def leftright(p, left, right, font='b'):
    if font == 'a':
        p.textln("{:<20}{:>12}".format(left, right))
    else:
        p.textln("{:<30}{:>12}".format(left, right))
```

Maybe we can think of a kind of API for this. At least it will help if someone wants to do this right now without using a cleaner solution (I think this is Python 3 only).

**Comment by praneshvrp1396** [comment]
Is there any solution available for question in any of the latest updates?

**Comment by patkan** [comment]
As you can see in this issue, the behavior of hardware tabs has been fixed in #238. You could therefore try issuing tabstops. Apart from that no other relevant changes have been implemented.

**Comment by VitorHugoOli** [comment]
I create this solution isn't perfect, cause utilize the max column length from profile, but works
``` python
def line_align_texts(p: printer.Serial, texts:List[str]):
    num_col = p.profile.get_columns(0)

    space = round((num_col - sum(len(i) for i in texts)) / (len(texts) - 1))
    text_build = ''.join(i + (" " * space) for i in texts[:-1])
    text_build += texts[-1]

    p.set(align='left', font='a')
    p.textln(text_build)
```
